### ISSUE TITLE
Provide an additional build target that isolates ol3-google-maps

### ISSUE THREAD

**Issue by laurentboileau** [mapgears/ol3-google-maps#142]
Since OpenLayers is now being compiled with ol3-google-maps, it might be interesting to offer an additional build target that only outputs the built ol3-google-maps library.

The current approach of bundling OL3 & ol3-google-maps requires that the Google Maps API be available when the bundled script is loaded in order to perform the necessary `ol.inherits` calls.

Our use case is similar to the one outlined in the following pull request: https://github.com/mapgears/ol3-google-maps/pull/31

> [...] the library had to be dynamically loaded after detecting if Google Maps was actually loaded, which was kind of a pain.

In our scenario, OpenLayers is always loaded. The Google Maps API, however, may not be. If it's needed, we'll dynamically load it (and ideally an isolated build of ol3-google-maps) after loading OL.

---

If offering such as build target is not possible, could the `ol.inherits` instructions be deferred to prior `olgm.OLGoogleMaps` being initialized?

Thank you.

**Comment by adube** [comment]
Hello @laurentboileau 

A) I don't think creating a build that only contains OLGM would be possible at this stage.

B) The idea of deferring the `ol.inherits` seems like a good approach, however I'm not sure how we could accomplish this.  Do you have suggestions on the matter?

C) One thing you could do is create a custom build of OLGM, one that would only include the components from OpenLayers it actually requires without exposing them, i.e. without exposing `ol.interaction.DragPan` for example.  Look at the following tutorial: http://openlayers.org/en/latest/doc/tutorials/custom-builds.html.  In it, it explains how to create a custom build of OpenLayers to have it include and expose only the components we want.  OLGM uses the same build tools.  You could try to do the same thing.  The resulted minimized .js file would be compatible alongside a build of OpenLayers only.  The only disadvantage of this approach is you end up with certain components in both builds, but OLGM is relatively small so this shouldn't be a too big of a problem.

I think you could try C) as it would not require any development.  If you do end up succeeding, it would know how you acheived it.  Otherwise, B) seems reasonable but requires work.

**Comment by laurentboileau** [comment]
Hello @adube 

Thank you for the thorough reply. The C) option seems like the best candidate. I'll try it and document my process. Thank you for proposing it.

I'm not too keen on option B) either. While it would address my issue, I don't see how it could be cleanly implemented. My first thought was to isolate the `ol.inherits` calls inside a function that would only be ran once by our app prior to initializing our required OLGM objects...

**Comment by laurentboileau** [comment]
Hello,

I tried option C), but ran into issues. OpenLayers objects created using my initially loaded OL3 wouldn't be recognized (`instanceof`) by the duplicated OL3 code in my isolated OLGM build. As a result, map initialization would fail. Running debug builds would also fail, because `goog.provide` would detect the duplicated `ol` namespace and would stop loading OLGM.

I defaulted back to option B), but also had issues. In my case, I moved the `ol.inherits` calls outside of the OLGM library and called them inside my code prior to initializing OLGM objects. This would unfortunately wipe the childClass prototype functions that were defined after their constructors in OLGM. To go through with this, I'd have to move the `ol.inherits` calls _as well as_ the childClass prototype functions defined in OLGM. At that point, I rejected option B).

Finally, I opted to conditionally load the Google Maps API _prior_ to loading the currently distributed OLGM build. Unfortunately, it isn't quite the solution to the problem I raised, but it's nonetheless the best way to handle this situation IMO.

Thank you for a fantastic library and your help in dealing with this issue.

Closing...
